In [2]:
# Importing the necessary packages
import pandas as pd
import numpy as np
from imblearn.over_sampling import SMOTE
from sklearn.ensemble import RandomForestClassifier

# To ensure all outputs from a cell are shown
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# To ensure all columns of a dataframe are shown
pd.set_option("display.max_columns", None)

In [3]:
# Reading in the data
df = pd.read_csv("train.csv")
df

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0          0        109             0.0             0.0             0.0   
1          1        109             0.0             0.0             0.0   
2          2        109             0.0             0.0             0.0   
3          3        109             0.0             0.0             0.0   
4          4        109             0.0             0.0             0.0   
...      ...        ...             ...             ...             ...   
69994  69994        109             0.0             0.0             0.0   
69995  69995        109             0.0             0.0             0.0   
69996  69996        109             0.0             0.0             0.0   
69997  69997        109             0.0             0.0             0.0   
69998  69998        109             0.0             0.0             0.0   

      last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
0                6/30/2014            7/31/2014            8/31/2014   31.277   
1                6/30/2014            7/31/2014            8/31/2014    0.000   
2                6/30/2014            7/31/2014            8/31/2014   60.806   
3                6/30/2014            7/31/2014            8/31/2014  156.362   
4                6/30/2014            7/31/2014            8/31/2014  240.708   
...                    ...                  ...                  ...      ...   
69994            6/30/2014            7/31/2014            8/31/2014   15.760   
69995            6/30/2014            7/31/2014            8/31/2014  160.083   
69996            6/30/2014            7/31/2014            8/31/2014  372.088   
69997            6/30/2014            7/31/2014            8/31/2014  238.575   
69998            6/30/2014            7/31/2014            8/31/2014  168.269   

        arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  offnet_mou_6  \
0       87.009    7.527        48.58       124.38         1.29         32.24   
1      122.787   42.953         0.00         0.00         0.00          0.00   
2      103.176    0.000         0.53        15.93         0.00         53.99   
3      205.260  111.095         7.26        16.01         0.00         68.76   
4      128.191  101.565        21.28         4.83         6.13         56.99   
...        ...      ...          ...          ...          ...           ...   
69994  410.924  329.136         0.00         7.36        10.93          0.00   
69995  289.129  265.772       116.54       196.46       232.63         49.53   
69996  258.374  279.782        77.13        68.44        78.44        335.54   
69997  245.414  145.062        14.01         7.64         6.71         30.34   
69998   42.815  167.961         0.00         0.00         0.00          0.00   

       offnet_mou_7  offnet_mou_8  roam_ic_mou_6  roam_ic_mou_7  \
0             96.68          2.33           0.00            0.0   
1             25.99         30.89           0.00            0.0   
2             82.05          0.00           0.00            0.0   
3             78.48         50.23           0.00            0.0   
4             38.11          9.63          53.64            0.0   
...             ...           ...            ...            ...   
69994        488.46        381.64          14.96            0.0   
69995         96.28         48.06           0.00            0.0   
69996        227.94        263.84           0.00            0.0   
69997         16.68         12.56          25.06            0.0   
69998          0.00          0.00           0.00            0.0   

       roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
0                0.0           0.00            0.0           0.00   
1                0.0           0.00            0.0           0.00   
2                0.0           0.00            0.0           0.00   
3                0.0           0.00            0.0           1.63   
4                0.0     

In [4]:
df.describe()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count  69999.000000    69999.0         69297.0         69297.0   
mean   34999.000000      109.0             0.0             0.0   
std    20207.115084        0.0             0.0             0.0   
min        0.000000      109.0             0.0             0.0   
25%    17499.500000      109.0             0.0             0.0   
50%    34999.000000      109.0             0.0             0.0   
75%    52498.500000      109.0             0.0             0.0   
max    69998.000000      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8   onnet_mou_6  \
count         69297.0  69999.000000  69999.000000  69999.000000  67231.000000   
mean              0.0    283.134365    278.185912    278.858826    133.153275   
std               0.0    334.213918    344.366927    351.924315    299.963093   
min               0.0  -2258.709000  -1289.715000   -945.808000      0.000000   
25%               0.0     93.581000     86.714000     84.095000      7.410000   
50%               0.0    197.484000    191.588000    192.234000     34.110000   
75%               0.0    370.791000    365.369500    369.909000    119.390000   
max               0.0  27731.088000  35145.834000  33543.624000   7376.710000   

        onnet_mou_7   onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
count  67312.000000  66296.000000  67231.000000  67312.000000  66296.000000   
mean     133.894438    132.978257    198.874771    197.153383    196.543577   
std      311.277193    311.896596    316.818355    322.482226    324.089234   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.675000      6.410000     34.860000     32.240000     31.575000   
50%       32.280000     32.100000     96.480000     91.885000     91.800000   
75%      115.837500    115.060000    232.990000    227.630000    229.345000   
max     8157.780000  10752.560000   8362.360000   7043.980000  14007.340000   

       roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  \
count   67231.000000   67312.000000   66296.000000   67231.000000   
mean        9.765435       7.014568       7.004892      14.186457   
std        57.374429      55.960985      53.408135      73.469261   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max      2850.980000    4155.830000    4169.810000    3775.110000   

       roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
count   67312.000000   66296.000000      67231.000000      67312.000000   
mean        9.842191       9.771783         46.904854         46.166503   
std        58.511894      64.618388        150.971758        154.739002   
min         0.000000       0.000000          0.000000          0.000000   
25%         0.000000       0.000000          1.660000          1.650000   
50%         0.000000       0.000000         11.910000         11.580000   
75%         0.000000       0.000000         40.740000         39.760000   
max      2812.040000    5337.040000       6431.330000       7400.660000   

       loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
count      66296.000000      67231.000000      67312.000000      66296.000000   
mean          45.686109         93.238231         90.799240         91.121447   
std          153.716880        162.046699        153.852597        152.997805   
min            0.000000          0.000000          0.000000          0.000000   
25%            1.610000          9.920000         10.090000          9.830000   
50%           11.740000         41.030000         40.170000         40.350000   
75%           39.895000        110.430000        107.540000        109.245000   
max        10752.560000       4696

The columns "circle_id", "loc_og_t2o_mou", "std_og_t2o_mou", "loc_ic_t2o_mou", "std_og_t2c_mou_6", "std_og_t2c_mou_7", "std_og_t2c_mou_8", "std_ic_t2o_mou_6", "std_ic_t2o_mou_7", "std_ic_t2o_mou_8" all have either 0 as the value or the same value for the whole dataset

In [7]:
# Check for columns with missing values
pd.set_option("display.max_rows", None)
df.isna().sum()
pd.set_option("display.max_rows", 10)

id                              0
circle_id                       0
loc_og_t2o_mou                702
std_og_t2o_mou                702
loc_ic_t2o_mou                702
last_date_of_month_6            0
last_date_of_month_7          399
last_date_of_month_8          733
arpu_6                          0
arpu_7                          0
arpu_8                          0
onnet_mou_6                  2768
onnet_mou_7                  2687
onnet_mou_8                  3703
offnet_mou_6                 2768
offnet_mou_7                 2687
offnet_mou_8                 3703
roam_ic_mou_6                2768
roam_ic_mou_7                2687
roam_ic_mou_8                3703
roam_og_mou_6                2768
roam_og_mou_7                2687
roam_og_mou_8                3703
loc_og_t2t_mou_6             2768
loc_og_t2t_mou_7             2687
loc_og_t2t_mou_8             3703
loc_og_t2m_mou_6             2768
loc_og_t2m_mou_7             2687
loc_og_t2m_mou_8             3703
loc_og_t2f_mou

The columns "fb_user_6", "fb_user_7", "fb_user_8", "night_pck_user_6", "night_pck_user_7", "night_pck_user_8", "arpu_3g_6", "arpu_3g_7", "arpu_3g_8", "arpu_2g_6", "arpu_2g_7", "arpu_2g_8","av_rech_amt_data_8", "av_rech_amt_data_7", "av_rech_amt_data_6", "count_rech_3g_8", "count_rech_3g_7", "count_rech_3g_6", "count_rech_2g_8", "count_rech_2g_7", "count_rech_2g_6", "max_rech_data_8", "max_rech_data_7", "max_rech_data_6", "total_rech_data_8", "total_rech_data_7", "total_rech_data_6", "date_of_last_rech_data_8", "date_of_last_rech_data_7", "date_of_last_rech_data_6", "last_date_of_month_6", "last_date_of_month_7", "last_date_of_month_8", all have >51,000 missing values out of 69,999 rows

In [64]:
# List of cols to drop with more than 50% missing values and same values
cols =["circle_id","loc_og_t2o_mou","std_og_t2o_mou","loc_ic_t2o_mou","std_og_t2c_mou_6","std_og_t2c_mou_7","std_og_t2c_mou_8","std_ic_t2o_mou_6","std_ic_t2o_mou_7","std_ic_t2o_mou_8",
      "fb_user_6","fb_user_7","fb_user_8","night_pck_user_6","night_pck_user_7","night_pck_user_8","arpu_3g_6","arpu_3g_7","arpu_3g_8","arpu_2g_6","arpu_2g_7","arpu_2g_8",
      "av_rech_amt_data_8","av_rech_amt_data_7","av_rech_amt_data_6","count_rech_3g_8","count_rech_3g_7","count_rech_3g_6","count_rech_2g_8","count_rech_2g_7","count_rech_2g_6",
      "max_rech_data_8","max_rech_data_7","max_rech_data_6","total_rech_data_8","total_rech_data_7","total_rech_data_6","date_of_last_rech_data_8","date_of_last_rech_data_7","date_of_last_rech_data_6",
      "last_date_of_month_6","last_date_of_month_7","last_date_of_month_8", "id", "date_of_last_rech_6","date_of_last_rech_7","date_of_last_rech_8"]

In [65]:
df= df.drop(cols, axis=1)

In [66]:
# Impute missing values with simple imputer, method is median
from sklearn.impute import SimpleImputer
imp_median = SimpleImputer(missing_values=np.nan, strategy='median')
imp_median.fit(df)
impute_data= pd.DataFrame(imp_median.transform(df))
impute_data.columns= df.columns

SimpleImputer(strategy='median')

In [85]:
# Perform feature selection using RFE
from sklearn.feature_selection import RFE
from sklearn.tree import DecisionTreeClassifier
X2=impute_data.drop(["churn_probability"], axis =1)
y2=impute_data["churn_probability"]
rfe = RFE(estimator=DecisionTreeClassifier(max_depth= 7), n_features_to_select=1)
rfe.fit(X2, y2)

RFE(estimator=DecisionTreeClassifier(max_depth=7), n_features_to_select=1)

In [86]:
# Get the ranks of the column
rank= pd.DataFrame()

rank['column']= X2.columns
rank['rfe_rank']= rfe.ranking_

In [87]:
# Sort the ranks in ascending
rank = rank.sort_values(by= 'rfe_rank')
cols = rank["column"].tolist()
columns_best= cols[0:30] + ["churn_probability"]

In [89]:
# Get the top 30 features
df2=impute_data[columns_best]

In [67]:
from fast_ml.model_development import train_valid_test_split
# Splitting the data into Training, validation and test sets
X_train3, y_train3, X_valid3, y_valid3, X_test3, y_test3 = train_valid_test_split(impute_data, target = "churn_probability", 
                                                                            train_size=0.7, valid_size=0.2, test_size=0.1, random_state=5)

In [ ]:
# Defining the model
RF2 = RandomForestClassifier()

# Hyperparameter tuning using rssandomizedsearch
modelRF2 = RandomizedSearchCV(estimator=RF2, param_distributions=param_grid, n_iter=10, cv=3, verbose=100, random_state=7, n_jobs=-1)

# Fitting the model
modelRF2.fit(X_train3, y_train3)

# Getting the best parameters
modelRF2.best_params_

In [93]:
# Training using the Random Forest Regressor
import timeit
# Creating the model
modelRF = RandomForestClassifier(n_estimators=500, max_depth=15, random_state=7, class_weight="balanced")

start = timeit.default_timer()

# Fitting the model
modelRF.fit(X_train3,y_train3)

stop = timeit.default_timer()

print('Time: ', stop - start)  

RandomForestClassifier(class_weight='balanced', max_depth=15, n_estimators=500,
                       random_state=7)

Time:  86.28100066600018


In [70]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
# Predicting Values for the training set
predsRFTrain = modelRF.predict(X_train3)
print("Training Confusion Matrix:")
confusion_matrix(y_train3, predsRFTrain)
print("Training Accuracy:")
accuracy_score(y_train3, predsRFTrain)

# Predicting the values for validation set
predsRF = modelRF.predict(X_valid3)

# Validation error
print("\nValidation Confusion Matrix:")
confusion_matrix(y_valid3, predsRF)
# Validation error
print("\nValidation Accuracy:")
accuracy_score(y_valid3, predsRF)

# Predicting the values for the test set
predsRF2 = modelRF.predict(X_test3)

print("\nTest Confusion Matrix:")
confusion_matrix(y_test3, predsRF2)
print("\nTest Accuracy:")
accuracy_score(y_test3, predsRF2)

Training Confusion Matrix:


array([[42922,  1047],
       [  161,  4869]])

Training Accuracy:


0.9753464356415437


Validation Confusion Matrix:


array([[12224,   342],
       [  443,   991]])


Validation Accuracy:


0.9439285714285715


Test Confusion Matrix:


array([[6127,  205],
       [ 203,  465]])


Test Accuracy:


0.9417142857142857

In [90]:
# Splitting the data into Training, validation and test sets
X_train4, y_train4, X_valid4, y_valid4, X_test4, y_test4 = train_valid_test_split(df2, target = "churn_probability", 
                                                                            train_size=0.7, valid_size=0.2, test_size=0.1, random_state=5)

In [94]:
# Training using the Random Forest Regressor
import timeit
# Creating the model
modelRF = RandomForestClassifier(n_estimators=500, max_depth=15, random_state=7, class_weight="balanced")

start = timeit.default_timer()
# Fitting the model
modelRF.fit(X_train4,y_train4)

stop = timeit.default_timer()

print('Time: ', stop - start)  

RandomForestClassifier(class_weight='balanced', max_depth=15, n_estimators=500,
                       random_state=7)

Time:  49.89910429200063


In [99]:
from sklearn.metrics import classification_report
# Predicting Values for the training set
predsRFTrain = modelRF.predict(X_train4)
print("Training Confusion Matrix:")
print(confusion_matrix(y_train4, predsRFTrain))
print("Training Accuracy:")
print(classification_report(y_train4, predsRFTrain))

# Predicting the values for validation set
predsRF = modelRF.predict(X_valid4)

# Validation error
print("\nValidation Confusion Matrix:")
print(confusion_matrix(y_valid4, predsRF))
# Validation error
print("\nValidation Accuracy:")
print(classification_report(y_valid4, predsRF))

# Predicting the values for the test set
predsRF2 = modelRF.predict(X_test4)

print("\nTest Confusion Matrix:")
print(confusion_matrix(y_test4, predsRF2))
print("\nTest Accuracy:")
print(classification_report(y_test4, predsRF2))

Training Confusion Matrix:
[[42943  1026]
 [  245  4785]]
Training Accuracy:
              precision    recall  f1-score   support

         0.0       0.99      0.98      0.99     43969
         1.0       0.82      0.95      0.88      5030

    accuracy                           0.97     48999
   macro avg       0.91      0.96      0.93     48999
weighted avg       0.98      0.97      0.97     48999


Validation Confusion Matrix:
[[12207   359]
 [  440   994]]

Validation Accuracy:
              precision    recall  f1-score   support

         0.0       0.97      0.97      0.97     12566
         1.0       0.73      0.69      0.71      1434

    accuracy                           0.94     14000
   macro avg       0.85      0.83      0.84     14000
weighted avg       0.94      0.94      0.94     14000


Test Confusion Matrix:
[[6116  216]
 [ 199  469]]

Test Accuracy:
              precision    recall  f1-score   support

         0.0       0.97      0.97      0.97      6332
         1